Product Collection: Hand picked products control in sidebar settings

imanish003 opened the pull request:
This PR introduces the ability to manually select specific products in the Product Collection block.
<img width="269" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a535113c-9c8c-4e1f-b7ff-eaaa1f2adade">

Changes include:
- Added `woocommerceHandPickedProducts` to the `ProductCollectionQuery` and `DEFAULT_FILTERS` in `constants.ts`.
- Created a new control file `hand-picked-products-control.tsx` which introduces a token field where the user can search for and select specific products.
- Included `HandPickedProductsControl` in the Product Collection block's inspector controls in `index.tsx`.
- Updated `ProductCollectionQuery` in `types.ts` to accommodate handpicked products.
- Updated the PHP `ProductCollection` class to handle the hand-picked products query parameters.

These changes allow users to hand-pick products to be displayed in the Product Collection block. This allows for greater customization of the products shown in the block.

Fixes #9722

### Testing

##### Simplest path

1. Add a “Product Collection” block to your page.
2. Go to the “Filters” within the Inspector Controls and add “Hand-picked Products”.
3. Start typing to select some products.
4. Ensure the preview is correctly updated as you narrow down your selection.
5. Click Publish.
6. Ensure the front-end is correct.

##### Complex path

1. Repeat steps 1–4 above.
2. Add additional filters (such as “On sale”).
3. Ensure the block only applies the subsequent filters to the subset of products you have selected.

Here is a quick demo:


https://github.com/woocommerce/woocommerce-blocks/assets/16707866/562efcdb-932c-41d7-8f73-572d73b9c54d



* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental


### Changelog

> Product Collection: Add a filter to allow merchants to hand-pick a subset of products.

imanish003 committed:
Add support for hand-picked products in Product Collection block

This commit introduces the ability to manually select specific products in the Product Collection block.

Changes include:
- Added `woocommerceHandPickedProducts` to the `ProductCollectionQuery` and `DEFAULT_FILTERS` in `constants.ts`.
- Created a new control file `hand-picked-products-control.tsx` which introduces a token field where the user can search for and select specific products.
- Included `HandPickedProductsControl` in the Product Collection block's inspector controls in `index.tsx`.
- Updated `ProductCollectionQuery` in `types.ts` to accommodate handpicked products.
- Updated the PHP `ProductCollection` class to handle the hand-picked products query parameters.

These changes allow users to hand-pick products to be displayed in the Product Collection block. This allows for greater customization of the products shown in the block.

imanish003 committed:
Merge branch 'trunk' into 9722-product-collection-filters-handpicked-products

thealexandrelara commented on assets/js/blocks/product-collection/inspector-controls/hand-picked-products-control.tsx:
```diff
@@ -0,0 +1,132 @@
+		return productsList.map( ( product ) => product.name );
+	}, [ productsList ] );
+
+	// Transform the token to a product name while displaying in the input field.
```
Maybe giving this function a more meaningful name such as `transformTokenIntoProductName` would allow you to remove this comment

imanish003 replied:
I see your point. By renaming the function to "transformTokenIntoProductName," the code would become more self-explanatory, making the comment redundant.

thealexandrelara reviewed (changes requested):
Code looks good to me, I just left a minor comment (not required to change, just a possible improvement)

I noticed that currently it is possible to add the same filter multiple times. I think that maybe we should filter out those that are already added to the input. For instance:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/20469356/e0f805cc-a48e-4855-8574-e20ff9cef92e)

In the image above, the Hoodie with Logo and Hoodie with Pocket are already duplicated, and if you search again for "Hoodie" they still appear in the list.

One issue that the duplication causes is that if I have two products that are the same, I can delete just one:


https://github.com/woocommerce/woocommerce-blocks/assets/20469356/d31343eb-335c-4e9c-a6cc-9d4ebb5fa992

imanish003 committed:
Enhance handling of hand-picked products

- A Set data structure is now used to store 'newHandPickedProducts' instead of an Array, which will help prevent duplicate entries.
- Additionally, the suggestions for products to be hand-picked now excludes already selected products, enhancing the user experience by avoiding redundancy in the suggestions list.
- Lastly, the function name 'displayTransform' has been changed to 'transformTokenIntoProductName' to more accurately reflect its purpose, which is to transform a token into a product name.

imanish003 commented:
Hey @thealexandrelara, great catch! The hand-picked filter is present in the Products beta block as well, and it seems to have the same issue. Thank you for the amazing PR feedback. I have implemented the suggested changes in commit [a3854ad](https://github.com/woocommerce/woocommerce-blocks/pull/9733/commits/a3854ad5c8ea2ee50774425842b49e3b56d7890a). Could you please take another look?

imanish003 committed:
Merge branch 'trunk' into 9722-product-collection-filters-handpicked-products

thealexandrelara reviewed (approved):
Great job! Everything is working as expected

✅ When selecting a product, the Product Collection reflects that on the Editor
✅ When selecting a product, the product is filtered out from the list, so I cannot add a duplicated entry
✅ When saving the page, the frontend is rendered accordingly to the selected filters

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9722-product-collection-filters-handpicked-products

imanish003 committed:
Update import & export of HandPickedProductsControl
